components/msg/background_hang_monitor.rs

```diff
@@ -2,50 +2,244 @@
+    regs.frame_ptr = state.__rbp as Address;
+}
+
+#[cfg(target_os = "windows")]
+#[allow(unsafe_code)]
+unsafe fn get_registers(thread_id: MonitoredThreadId, _regs: &mut Registers, _success: &mut bool) {
+    // TODO: use winapi::um::processthreadsapi::GetThreadContext
+    // and populate registers using the context.
+    // See https://dxr.mozilla.org/mozilla-central/source/tools/profiler/core/platform-win32.cpp#129
+}
+
+#[cfg(any(target_os = "android", target_os = "linux"))]
+#[allow(unsafe_code)]
+unsafe fn get_registers(_thread_id: MonitoredThreadId, _regs: &mut Registers, _success: &mut bool) {
     loop {
-        if BACKTRACE_READY.load(Ordering::SeqCst) {
+        if SAMPLEE_CONTEXT_READY.load(Ordering::SeqCst) {
             break;
         }
         thread::yield_now();
     }
-    let (thread_id, trace) = BACKTRACE.take().unwrap();
-    assert_eq!(thread_id, monitored.thread_id);
-    BACKTRACE_READY.store(false, Ordering::SeqCst);
-    Some(trace)
+    let (_thread_id, _context) = SAMPLEE_CONTEXT.take().unwrap();
+    // TODO: populate registers from context.
+    // See https://dxr.mozilla.org/mozilla-central/source/tools/profiler/core/platform-linux-android.cpp#85
+}
+
+#[cfg(target_os = "macos")]
+#[allow(unsafe_code)]
+unsafe fn resume_thread(thread_id: MonitoredThreadId, success: &mut bool) {
+    check_kern_return(mach::thread_act::thread_resume(thread_id), success);
 }
 
 #[cfg(target_os = "windows")]
-unsafe fn get_backtrace_from_monitored_component(monitored: &MonitoredComponent) -> Option<Backtrace> {
-    None
+#[allow(unsafe_code)]
+unsafe fn resume_thread(thread_id: MonitoredThreadId, _success: &mut bool) {
+    // TODO: use winapi::um::processthreadsapi::ResumeThread
+}
+
+#[cfg(any(target_os = "android", target_os = "linux"))]
+fn resume_thread(_thread_id: MonitoredThreadId, _success: &mut bool) {
+    SAMPLEE_MAY_RESUME.store(true, Ordering::SeqCst);
+}
+
+#[allow(unsafe_code)]
+unsafe fn frame_pointer_stack_walk(regs: &mut Registers, native_stack: &mut NativeStack) {
+    // Note: this will only work with code build with debug flags,
+    // or RUSTFLAGS=-Cforce-frame-pointers=yes
+    let pc = regs.instruction_ptr as *mut std::ffi::c_void;
+    let stack = regs.stack_ptr as *mut std::ffi::c_void;
+    native_stack.process_register(pc, stack);
+    let mut current = regs.frame_ptr as *mut *mut std::ffi::c_void;
+    let mut frame_count = 1;
+    while !current.is_null() && frame_count < MAX_NATIVE_FRAMES {
+        let next = *current as *mut *mut std::ffi::c_void;
+        // NOTE: assuming i386 or powerpc32 linux.
+        // TODO: ppc mac or powerpc64 linux require add(2) and add(3) below.
+        let pc = current.add(1);
+        let stack = current.add(2);
+        native_stack.process_register(*pc, *stack);
+        frame_count = frame_count + 1;
+        current = next;
+    }
+}
+
+fn symbolize_backtrace(native_stack: &mut NativeStack, hang_profile: &mut Option<HangProfile>) {
```

Comment:
Same comment about returning a value instead of accepting a mutable argument. Additionally, since we only return Some, we can drop the Option.